reference/programming-model.md

```diff
@@ -33,80 +64,243 @@ The [@pulumi/pulumi] package is the core library for working with the Pulumi pla
 - External cloud platform resources ([pulumi.CustomResource])
 - Components defined entirely in JavaScript ([pulumi.ComponentResource])
 
-Dependencies between resources are encoded with [pulumi.Output].
+Dependencies between resources are encoded with [pulumi.Output](#outputs).
 
 This package also provides the following helpers:
 - [pulumi.getStack] to get information about the current stack 
 - [pulumi.log] for logging deployment information
-- [pulumi.runtime.serializeFunctionAsync] for turning JavaScript callbacks into data which can be [used as application code](#runtime).
+- [pulumi.runtime.serializeFunction] for turning JavaScript callbacks into data which can be [used as application code](#runtime).
 
 ## Creating resources {#resources}
 
-A resource is created via `new Resource(name)` in JavaScript. All resources must have a name, which must be unique in the Pulumi program.
+A resource is created via `new Resource(name, args)` in JavaScript. All resources must have a name, which must be unique in the Pulumi program.
 
-All resource constructors take the following additional properties. 
-- `dependencies` - a list of explicit resource dependencies
+All resource constructors also accept a third argument which can provide the following additional properties. 
+- `dependsOn` - a list of explicit resource dependencies
 - `protect` - whether to mark a resource as protected. A protected resource cannot be deleted directly: first you must set `protect: false` and run `pulumi update`. Then, the resource can be deleted, either by removing the line of code or by running `pulumi destroy`.
 - `parent` - optional parent for the resource. See [Components](#components).
 - `provider` - optional provider for the resource. See [Providers](#providers).
 
 ## Resource outputs {#outputs}
 
-<!-- TODO: add direct anchor link to `apply` once #333 is fixed -->
+Outputs are a key part of how Pulumi tracks dependencies between resources.  Because the values of Outputs are not available until resources are created, these are represented using a special [Output] type which internally represents two things:
+1. An eventually available value of the output
+2. The depdency on the source(s) of the output value
+
+The output properties of all resource objects in Pulumi have type [`Output`][pulumi.Output]. Resource inputs have type [`Input`][pulumi.Input], which accepts either a raw value or an output from another resource. This allows dependencies to be inferred, including ensuring that resources are not created or updatred until all their dependencies are availabe and up to date.  
```

Comment:
Do you want to mention that `Input` can also be a `Promise`?

Also, in general, I often draw an analogy between `Output`s and `Promise`s -- e.g., "`Output`s are just `Promise`s that carry dependency information." Not sure if that analogy makes sense to state explicitly here.